modules/shell/output.go

```diff
@@ -0,0 +1,88 @@
+package shell
+
+import (
+	"strings"
+	"sync"
+)
+
+// output contains the output after runnig a command.
+type output struct {
+	stdout *outputStream
+	stderr *outputStream
+	// merged contains stdout  and stderr merged into one stream.
+	merged *merged
+}
+
+func newOutput() *output {
+	m := new(merged)
+	return &output{
+		merged: m,
+		stdout: &outputStream{
+			merged: m,
+		},
+		stderr: &outputStream{
+			merged: m,
+		},
+	}
+}
+
+func (o *output) Stdout() string {
+	if o == nil {
+		return ""
+	}
+
+	return o.stdout.String()
```

Comment:
Do you need to check if `o.stdout == nil` as well?

Reply:
No, as even if `o.stdout` is nil, we can call `*outputStream.String()` with a nil pointer.
The actual `String()` methods check if the pointer is nil and just return an empty string in this case

(I actually noticed that I missed `*outputStream.String`, but I did this with `*merged.String` already. Now fixed and implemented with outputStream too.)